### DISCUSSION TITLE
Could `if` statement makes sense here?

### DISCUSSION THREAD

**Discussion by manuel-rubio** [gleam-lang/gleam#2158] [Ideas & suggestions]
I was reading code like this:
```gleam
  case list.length(events) >= tracker.limit {
    True -> Error(TooIntense)
    False -> Ok(IntensityTracker(..tracker, events: events))
  }
```
Because I came from Erlang and Elixir, I could consider something that's clear to keep it as Erlanger but maybe it could be simplified (my alchemist is yelling about it! :-D ):
```gleam
  if list.length(events) >= tracker.limit {
    Error(TooIntense)
  } else {
    Ok(IntensityTracker(..tracker, events: events))
  }
```
Of course, it should be useful and used for Bool evaluations, what do you think?

**Comment by inoas** [comment]
Personally I'd like an if/else that:
- is an expression obviously
- enforces explicit else
- only takes boolean (no boolish) expressions
- short cuts
- does not have any elseif feature

```gleam
if foo() && bar() && quux() {
  True
} else {
  False
}
```

...goes into `else` after `foo()` returned `False`.

That being said, except for reserving if and else as keywords this can be added quite easily down the line, and the gains are small I guess, so IMHO this is quite low prio.

**Comment by lpil** [comment]
I think `if/else` could be a possibility in future. Let's see how we get with `use` to start as it can do the same thing.

**Reply by inoas** [reply]
use can short-circuit? (or rather I can built short circuit logic in a terse way with use?)

**Reply by lpil** [reply]
Yes, check out `bool.guard`

**Reply by inoas** [reply]
I meant this:

> /// Note that unlike in procedural languages the `return` value is evaluated
> /// even when the predicate is `False`, so it is advisable not to perform
> /// expensive computation there.

**Comment by inoas** [comment]
I think having if/else helps with strangeness budget. We could even support if/elseif/else by throwing a specific compiler error that says to use case?

**Reply by tryoxiss** [reply]
This! I think that the case is clearer and adding multiple ways to do the same thing would go against the languages stated goals, but having an error that shows how to turn it into gleam syntax (with `case`) would be helpful for new people without going against the goals (simillar to how `panic(msg)` gives an error and says to use `panic as msg`).

**Reply by lpil** [reply]
Clever! Could you make an issue for that, please?

**Reply by tryoxiss** [reply]
Sure, created #3572 !

**Comment by timjs** [comment]
What about adding a multi-way conditional expression, like [`if` in Erlang](https://www.erlang.org/doc/reference_manual/expressions.html#if), [`cond` in Elixir](https://elixir-lang.org/getting-started/case-cond-and-if.html#cond) and [multi-way `if` in Haskell](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/multiway_if.html)?

Currently you'd write:
```gleam
case c1 {
  True -> e1
  False -> e0
}
```
 (38 characters).
With an if-then-else this would become:
```fsharp
if c1 {
  e1
} else {
  e0
}
```
(28 characters).
I'd say something like:
```fsharp
when {
  c1 -> e1
  else -> e0
}
```
(32 characters) would be readable and easily expandable.

Take an example with more boolean tests:
```gleam
case c1 {
  True -> e1
  False -> case c2 {
    True -> e2
    False -> e0
  }
}
```
(80 characters).
With if-then-else this would become:
```fsharp
if c1 {
  e1
} else { if c2 {
  e2
} else {
  e0
} }
```
(52 characters when formatted without an extra indent).
With multi-way branching:
```fsharp
when {
  c1 -> e1
  c2 -> e2
  else -> e0
}
```
(42 characters).

This would:
* be an expression obviously
* enforce explicit else
* only take boolean (no boolish) expressions
* short cut
* take multiple branches

Another benefit: using this multi-way if, the right-right-hand side of the arrows can be single expressions in most cases, and block expressions only when needed, like in Gleam's `case` expression. The if-then-else forces block expressions for both branches.

**Reply by inoas** [reply]
I like the multi way / cond approach except else feeling syntactically strange, here are a few variants on when.

```rust
when {
  c1 -> e1
  c2 -> e2
  e0
}
```

```rust
when {
  c1 -> e1
  c2 -> e2
  else e0
}
```

```rust
when {
  c1 -> e1
  c2 -> e2
  or e0
}
```


```rust
when {
  c1 -> e1
  c2 -> e2
  default e0
}
```

```rust
when {
  c1 -> e1
  c2 -> e2
} or e0
```

```rust
when {
  c1 -> e1
  c2 -> e2
} else e0
```

**Reply by lpil** [reply]
While I think this syntax is in many ways nicer, I don't think it's enough of an advantage to spend some of our strangeness budget on. People are very familiar with if/else so I think if we have that feature we should use the overwhelmingly common syntax.

**Reply by inoas** [reply]
Can't we just reserve some words like when, where, else, or, cond and take a look later?

**Reply by timjs** [reply]
Yeah, if-then-else is really common! But I just don't think it fits Gleam 🫤

Maybe reserving `if` and adding a hint in the error message would already help a lot?

**Reply by inoas** [reply]
`if` is already reserved, unlike `when`, `where`, `cond`, `else`, `or`

**Reply by timjs** [reply]
I realised I didn't explain why I think if-then-else doesn't fit Gleam. Let me try to elaborate.

Gleam is an _expression based_ language. One possible kind of expression is a block expression, which evaluates to the last expression in the block. This is consistent throughout Gleam's language constructs:

```gleam
let x = 1 * 2  // short
let y = {      // long
  let a = f()
  Some(a)
}
let z = case y {
  None -> 42   // short
  Some(v) -> { // long
    use w <- g()
    v * w
  }
}
```

Because of this, I think an `if c { e1 } else { e2 }` syntax doesn't fit Gleam because, as I mentioned earlier, it _forces_ you to always use a block, even for short expressions:
```gleam
let e = if is_some(y) { // could be short...
  42
} else {
  use w <- g()
  37 * w
}
```
Obviously, YMMV, but I don't like this inconsistency.

One could mitigate this, by introducing an `if c then e1 else e2` expression like StandardML/OCaml/F# like this:
```gleam
let e = if is_some(y)
  then 42
  else {
    use w <- g()
    37 * w
  }
```
Problem is, Gleam uses curly braced syntax C-style syntax, where a `then` keyword is not common. So doing above, would trip over users coming from JavaScript, Rust, etc. (and even Python, which doesn't have a `then` either). Combining (imperative) curly-braced syntax with (functional) expression based semantics is a challenge...

That's why I think something like `cond` as found in Elixir/Scheme/Racket fits a functional expression based language like Gleam better. It also resembles the already used `case` construct, but then evaluating predicates top-to-bottom instead of pattern matching top-to-bottom.
```gleam
let e = when {
  is_some(y) -> 42 // short
  True -> {        // long
    use w <- g()
    37 * w
  }
}
```
(With any keyword/underscore/... signalling the last branch is the default branch and is always evaluated in in the end.)

But, in the end, I can understand that an argument "this is more natural and recognisable for the intended user base" overrules the argument "this fits the value that everything in the language is an expressions, and longer expressions are grouped in a block".

**Reply by inoas** [reply]
I don't see the issue with `if condition {body} else condition {body}`

```rust
if {a && b} {
  c
} else {
  d
}
```
... looks fine to me.
However I also like cond/when as it can be a very concise and powerful multi way switch. I want to see that eventually tbh.

**Reply by bentomas** [reply]
I love this `when` syntax.  It feels like it matches `case` syntax but in a usefully different way, is very flexible, and can be super concise or more verbose.  I do wonder if you avoid some of the "strangeness" if you just call it `if`...

```
if {
  c1 -> e1
  c2 -> e2
  else -> e0
}

if {
  c1 -> {
    e1
  }
  else -> e0
}
```
(this is starting to look a lot like erlang's `if` but not limited to guards.  could just skip the `else` and go all in on that syntax and use `True`, at least it has precedent and would have some familiarity)

one thing i like about this over an `else if` is that it makes for cleaner diffs in version control when adding new clauses.  

the other thing i don't like about `if condition {body} else condition {body}` is that in gleam the `else` wouldn't be optional, right? since it is an expression, doesn't it always have to return a value?  so in `let x = if cond {body}` nothing would be returned if the `cond` is false. yet the syntax would sure imply that the else is optional.  at a minimum, it is optional in other languages, so new users would get tripped up on that.

**Reply by inoas** [reply]
I really like the idea of support if like that, except for the curleys

i'd love something like this:

```rust
if c1 -> expression1
   c2 -> expression2
 else -> expression0
```
or if the body is longer

```
if c1 ->
     expressionexpressionexpressionexpressionexpression1
   c2 -> e2
 else -> e0
```

**Comment by human154** [comment]
Isn't this currently possible and still more concise than `use` and bool.guard?

```gleam
  case any_irrelevant_value {
    _ if x < 10 -> io.println("small")
    _ if x < 100 -> io.println("medium")
    _ -> io.println("large")
  }
```

we could allow this at least:
```gleam
  case _ {
    _ if x < 10 -> io.println("small")
    _ if x < 100 -> io.println("medium")
    _ -> io.println("large")
  }
```
Allowing the use of the underscore (as shown above) in place of needing to supply an irrelevant value... it's a consistent syntax.

**Reply by lpil** [reply]
That works, but guards only permit a very limited set of expressions in them. You can't call a function, for example.

**Reply by human154** [reply]
Why are guards limited? ...its not obvious to me why this would be handled differently from an `if else` branching structure.

**Reply by lpil** [reply]
The virtual machine restricts guards for performance reasons.

**Comment by human154** [comment]
Dropping more syntax:

**Comment by human154** [comment]
```elm
  case {
    if x < 10 -> io.println("small")
    if x < 100 -> io.println("medium")
    _ -> io.println("large")
  }
```

**Reply by inoas** [reply]
I don't like this because I'd assume I can use guards on the right side too

**Comment by human154** [comment]
```elm
  case {
    x < 10 -> io.println("small")
    x < 100 -> io.println("medium")
    _ -> io.println("large")
  }
```
This would be roughly equivalent to kotlin's `when`:

```kotlin
    // Kotlin
    // "when" can be used as an alternative to "if-else if" chains.
    val i = 10
    when {
        i < 7 -> println("first block")
        fooString.startsWith("hello") -> println("second block")
        else -> println("else block")
    }

    // "when" can be used with an argument.
    when (i) {
        0, 21 -> println("0 or 21")
        in 1..20 -> println("in the range 1 to 20")
        else -> println("none of the above")
    }
```
I think these possibilities are cleaner than  `if..else if..else` and more consistent with existing Gleam syntax.  There is less syntactic keyword noise, conditions are placed front and center, and the conditions get vertical alignment (which they don't with `if then else` because `if` is shorter than `else if`).

**Comment by human154** [comment]
This is currently valid syntax:

```gleam
case x < 10 {
    True -> "small"
    False -> 
case x < 100 {
    True -> "medium"
    False -> 
case x < 1000 {
    True -> "large"
    False -> "extra large"
}}}
```
It reads nicely.

**Reply by inoas** [reply]
if it was formatted like this maybe >:D

**Comment by NNBnh** [comment]
I love Gleam dedication for minimalism to only have one type of flow control: `case`. But I just wandering and try to play devil's advocate, so I make If-Elif-Else in Gleam:

```gleam
import gleam/option.{type Option, None, Some}
import gleam/io

fn if_then(requirement: Bool, consequence: fn() -> a) -> Option(a) {
  case requirement {
    True -> Some(consequence())
    False -> None
  }
}

fn elif_then(value: Option(a), requirement: Bool, consequence: fn() -> a) -> Option(a) {
  case value, requirement {
    Some(value), _ -> Some(value)
    None, True -> Some(consequence())
    _, _ -> None
  }
}

fn else_then(value: Option(a), consequence: fn() -> a) -> Option(a) {
  case value {
    Some(value) -> Some(value)
    None -> Some(consequence())
  }
}

pub fn main() {
  if_then(False, fn() {
    io.println("Foo")
  }) |> elif_then(True, fn() {
    io.println("Bar")
  }) |> else_then(fn() {
    io.println("Baz")
  })
}
```

**Reply by human154** [reply]
@NNBnh that looks like something that could be translated into `use <-` syntax

Edit: link to bool.guard as per Ipil's comment below.
https://hexdocs.pm/gleam_stdlib/gleam/bool.html#guard

**Reply by lpil** [reply]
Check out bool.guard

**Reply by NNBnh** [reply]
I agree, `case` and `bool.guard` are 99.9% the best way to get the job done in FP language like Gleam. If-Else just doesn't useful much.

**Reply by manuel-rubio** [reply]
I have two objections about `bool.guard`:

![IMG_0488](https://github.com/user-attachments/assets/5e00ceac-e748-47a5-b8f8-acc1b933a82b)

First, the note should say that if you want to use side-effects or something complex, you should use lazy_guard instead (it's always good idea give a solution instead of only a limitation).

Second, the example:

```gleam
let name = ""
use <- guard(when: name == "", return: "Welcome!")
"Hello, " <> name
// -> "Welcome!"
```

Where we changed the value of `name`?

**Reply by lpil** [reply]
Ah that's a mistake in the documentation. I'll get that fixed.

The other bit is due to this function being older than the lazy one so when this was written there wasn't one to recommend.

**Reply by lpil** [reply]
I've read it again while not in the middle of a long day of travelling, and this time I'm reading it correctly 😁 

The example is correct. The variable `name` is never evaluated in that example.

**Reply by manuel-rubio** [reply]
oh! sorry, I re-read it again, you're right, both examples are right, sorry again.

**Reply by danielytics** [reply]
What about adding a functional ternary to go alongside guard?

`bool.if_else(when: Bool, true: a, false: a) -> a`

It would have a similar use to guard except the else (false) isn’t lazy like in guard. 

I personally already use such a function for cases like `if_else(value > threshold, use_value_a, use_value_b)`

There is also some prior art for such a function, so it wouldn’t really take from the weirdness budget:

* R, Julia, and MATLAB have `ifelse(c, a, b)`
* Pine Script has `iff(c, a, b)`

The implementation would simply be:

```gleam
pub fn if_else(when condition: Bool, true true_val: a, false false_val: a) -> a {
  case condition {
    True -> true_val
    False -> false_val
  }
}
```

**Reply by lpil** [reply]
What's the advantage of this over a case expression?

**Reply by danielytics** [reply]
That’s a question for this entire discussion, right? I mean, the question about adding if-else at all is whether there’s an advantage over case. I think in general, case is the more flexible and useful tool and adding a functional-ternary isn’t necessary, however guard exists and (in my opinion, others may disagree) a function ternary makes some code terser and more readable.

Where I personally use it is code like this:

```gleam
let healing_rate = if_else(healing_buff, buffed_healing_rate, passive_healing_rate)
let stamina_recovery_rate = if_else(stamina_buff, buffed_stamina_rate, passive_stamina_rate)
```
Where many such selections (for different rates and values —and types) are made in succession, this is easier to read and maintain than using case:

```gleam
let healing_rate = case healing_buff {
  True -> buffed_healing_rate
  False -> passive_healing_rate
}
let stamina_recovery_rate = case stamina_buff {
  True -> buffed_stamina_rate
  False -> passive_stamina_rate
}
```

I mean, this is a contrived example and there are other better ways this could be done, and you could certainly make a case that this explicit code is more readable (in this example I think case us easier to understand at a glance, even if more verbose…), but my point is simply that it quickly becomes a lot of repeated code and at some point longer code is harder to read simply because there’s a lot of it, even if each individual part is simpler in isolation.

In my case, I made a little helper function that does this. It’s not a big deal. 

I see it the same way I see bool.to_int or even bool.guard: trivial technically-unnecessary functions (it’s two seconds to define them yourself, if needed), but still nice to have provided in the standard lib.

I don’t feel strongly about it either way, though. I’m suggesting it only because it seems to fit with guard and other languages provide it.

**Reply by lpil** [reply]
I find the function one harder to read, but also they're different. The function one performs worse as it evaluate both arguments.

If it was like this it would be closer, but there's still some overhead

```gleam
let healing_rate =
  if_else(healing_buff, fn() { buffed_healing_rate }, fn() {
    passive_healing_rate
  })
let stamina_recovery_rate =
  if_else(stamina_buff, fn() { buffed_stamina_rate }, fn() {
    passive_stamina_rate
  })
```

Note I have run the formatter on this code.

I don't think this is any easier to understand, and one of Gleam's design goals is to not have multiple ways to do the same thing. I don't see any advantages to this presently.

> bool.to_int

Good point. I will deprecate this now.

**Reply by danielytics** [reply]
if think in situations where you want them to be lazily evaluated, case comes up the clear winner — once you wrap each argument in a function, case is cleaner and shorter. The benefit only exists when you’re switching between two simple values.

Like I said, the example is contrived, but I do find a helper function like this made my code more readable because it let me create a series of one-liners. But also, I’m not sure it adds enough value to include because I think it’s uses are very limited and it’s so easy to just create if desired, and the only reason I suggested it is because other languages deemed it useful to include and it seemed to fit with the discussion and similarity to guard. I won’t be upset if you say “ok, but no thanks”.

> Good point. I will deprecate this now.

I didn’t mean it shouldn’t exist 😅 I mentioned it as an example of existing trivial functions that might still be beneficial to have included, sometimes. I haven’t looked very hard, but I assume there are many trivial functions in stdlib and I think being trivial isn’t enough to say they don’t have a place there.

> one of Gleam's design goals is to not have multiple ways to do the same thing

Yes, absolutely. There are already cases like guard where there are multiple ways to do the same thing, but they have a much more general set of use cases and clearer value. So, after hearing your thoughts and thinking about it more, I think gleam shouldn’t include this and personally I don’t think an if-else expression really fits either. Also it does seem like the prior art is perhaps solving a different problem (In R and presumably Julia and MATLAB, it seems it can be applied over a sequence of conditions, like a list.map, not only single conditions).

So, sorry for the noise 😅
